COMPATIBILITY.md

```diff
@@ -0,0 +1,109 @@
+release (because Bazel compatibility policy doesn't allow us to make that
+promise, and some new features of `rules_rust` require new Bazel features that
+are only available in Bazel rolling releases). `rules_rust` will aim that new
+features available to the current Bazel rolling release will be available to the
+next Bazel LTS release at latest.
+
+Whenever there is a new Bazel LTS release, all releases of `rules_rust` in the
+first 3 months will support both Bazel LTS versions unless Bazel doesn't allow this.
+
+## What host platforms are supported?
+
+The only platform subject to backwards compatibility policy is
+`x86_64-unknown-linux-gnu`, as the only fully supported platform. Process for
+moving a best effort platform to a supported platform is consensus-based.
+
+## What are the stable APIs of `rules_rust`?
+
+`//rust:defs.bzl` is subject to the backwards compatibility policy. That means
+that anything directly accessible from this file is considered stable.
+
+`//rust/private/…` is not subject to the backwards compatibility policy. Content
+of this package is an implementation detail.
+
+`//cargo:cargo_build_script.bzl` is subject to the backwards compatibility policy.
+
+`//cargo`, `//util`, `//tools`, `//test`, `//examples`, `//bindgen`, `//proto`,
+`//wasm_bindgen` and any packages not mentioned by this document are by default
+not subject to the backwards compatibility policy.
+
+Experimental build settings are not subject to the backward compatibility
+policy. They should be added to `//rust:experimental.bzl`.
+
+Incompatible build settings are subject to the backward compatibility policy,
+meaning the behavior of the flag cannot change in a backwards incompatible way.
+They should be added to `//rust:incompatible.bzl`.
+
+Bug fixes are not a breaking change by default. We'll use Common Sense (and we
+will pull in more maintainers and the community to discuss) if we see a certain
+bug fix is controversial. Incompatible changes to
+`//cargo:cargo_build_script.bzl` that make `cargo_build_script` more accurately
+follow cargo's behaviour are considered bug fixes.
+
+## How to make a backwards incompatible change?
+
+1. Create a GitHub issue (example:
+[Split rust_library into smaller rules#591](https://github.com/bazelbuild/rules_rust/issues/591)).
+2. Describe the change, motivation for the change, provide migration
+instructions/tooling.
+3. Add a build setting into `//rust:incompatible.bzl` that removes the old
+behavior (whenever possible) or changes the current behavior (when just
+removing the old behavior is not possible). Ideally, users should not need to
+manually flip incompatible flags.
+4. Mention the link to the GitHub issue in error messages. Do not add a
+deprecation warning (warnings make the deprecation visible to every user
+building a project, not only to the maintainers of the project or the rules).
+5. Mention the issue in the `CHANGELOG` file.
```

Comment:
Are users expected to update this file? I'd like to avoid that requirement if possible or automate it away somehow.

Reply:
If by users you mean contributors to rules_rust who introduce incompatible changes, then yes I thought so.

If you have plans to automate the release notes generation process and maybe scrape commit descriptions for some RELNOTES tag or what not, we can always update the text of the policy. Or do you want to update the text right now?

Reply:
> If by users you mean contributors to rules_rust who introduce incompatible changes, then yes I thought so.

Yeah, that's what I meant.

I suppose it's fine that users introducing those changes would need to update the `CHANGELOG`. For some reason I had it in my head that the changelog would contain "notable" changes in a release which are not limited to incompatible changes. If `CHANGELOG` requires that then I'd like to dive into the details there a bit more but if it's only required for incompatibility changes, then I'm a happy camper 😄

Reply:
In my mind, CHANGELOG contains both new features, notable bug fixes, new incompatible changes, or flipped incompatible changes, or anything useful. So would you like to dive into the details here in this issue?

Reply:
No, I can agree to making that part of the `CHANGELOG`. My only concern is about making sure that's effortlessly updated. I don't want to run into situations where a change that should have been in `CHANGELOG` wasn't added and now it's up to whoever is making a release to check that. Nor do I want users to feel like they need to add that for any change they make. A manually managed `CHANGELOG` file, in my experience, has added lots of churn across PRs and can turn into a time sink or a reason not to perform a release. Are there other projects that can more-or-less automate updating that file? Maybe something maintainers can comment on a ticket to have that updated pre-merge? This can also be a separate thread.